text/0000-link-alias.md

```diff
@@ -0,0 +1,316 @@
+#[link(name = "bar", alias = "foo")]   // explicitly aliased as `foo`
+```
+
+```
+rustc -l foo            # implicit alias is called "foo"
+rustc -l alias=foo=bar  # "bar" is explicitly aliased as "foo"
+```
+
+A `#[link(alias = "...")]` annotation is required to resolve to some native
+library, and an error will be generated if the alias has not been introduced.
+The culmination of this is now the compiler can connect an alias directive to a
+native library to understand that the symbols in the `extern` block are
+contained in that native library.
+
+#### Alias examples
+
+Some example usage of introducing aliases and then using them looks like:
+
+```rust
+// compiled with: rustc -l alias=a1=d1
+
+#[link(name = "lib1")]
+extern {}
+
+// aliases the library defined above
+#[link(alias = "lib1")]
+extern {}
+
+// aliases the library `d1` on the command line
+#[link(alias = "a1")]
+extern {}
+
+// also aliases the library `d1` on the command line
+#[link(alias = "d1")]
+extern {}
+
+// introduce the alias `a2` for `lib2
+#[link(name = "lib2", alias = "a2")]
+extern {}
+
+// aliases the library `lib2` above
+#[link(alias = "a2")]
+extern {}
+```
+
+### Treatment of dll{import,export}
+
+As a recap, let's take a look at today's treatment of dll{import,export} in the
+compiler with respect to native libraries. Note that the terms "functions" and
+"static" here refer to those defined in native libraries (e.g. connected via
+FFI). Currently dllimport is only applied when a static is referenced through an
+external crate. References to locally declared statics or functions in any
+circumstance never have dllimport applied. The compiler also currently has an
+unstable `#[linked_from]` attribute which it leverages to apply the dllexport
+annotation to statically linked native libraries, but otherwise dllexport is
+never applied. This treatment is incorrect in [a number of ways][issue], and is
+a strong motivating factor for this RFC!
+
+[issue]: https://github.com/rust-lang/rust/issues/27438
+
+Armed with `alias` directives, the compiler is now able to handle
+dllimport/dllexport correctly in all cases for native libraries. The dllimport
+attribute will be applied to all symbols in an `extern` block if that block has
+any linkage directive indicating that the symbols are linked via a dynamic
+library. (e.g. following alias pointers to their concrete linkage directives).
+Similarly, dllexport will only be applied to a block of symbols if a directive
+indicates that they're linked statically.
+
+Example application of dllexport/dllimport looks like:
+
+```rust
+// compiled with: rustc -l l1
+
+#[link(alias = "l1")]
+extern {
+    // dllimport applied, dllexport not applied
+}
+
+#[link(name = "l2")]
+extern {
+    // dllimport applied, dllexport not applied
+}
+
+#[link(name = "l3", kind = "static")]
+extern {
+    // dllimport not appplied, dllexport applied if linked staticaly to dylib
```

Comment:
Pointing out that `dllexport` is only applied when linked statically to a dylib seems a bit redundant as the kind is already specified as static so we know it will be linked statically and `dllexport` only matters when creating a dylib anyway.

Also it is important to note that it should only be `dllexport`ed from the first immediate dylib that is created. So if I create `a.dll` which depends on `b.dll` which statically depends on a static native library `c.lib`, then symbols from `c.lib` which end up as part of the interface of `b.dll` should be `dllexport`ed from `b.dll` but if they then end up in the interface of `a.dll` they should __not__ be `dllexport`ed since the `c` symbols don't exist in `a.dll`. Instead consumers of `a.dll` need to link to `b.dll`'s import library and get the relevant `c` symbols from there.